### PR TITLE
feat: opts.genid  for v4

### PR DESCRIPTION
**Pull request by shaoshuai0102**

### CODE REVIEW

**Comment by shaoshuai0102** [comment]
ping @dead-horse

File: lib/context.js [hunk 1]
```diff
@@ -177,7 +177,7 @@ class ContextSession {
 
   create(val, externalKey) {
     debug('create session with val: %j externalKey: %s', val, externalKey);
-    if (this.store) this.externalKey = externalKey || uid.sync(24);
+    if (this.store) this.externalKey = externalKey || this.opts.generateKey && this.opts.generateKey(this.ctx) || uid.sync(24);
```
Comment (dead-horse) [inline lib/context.js]: https://github.com/expressjs/session#genid use the same name as express session

Comment (dead-horse) [inline lib/context.js]: need add this option to README

**Comment by shaoshuai0102** [comment]
done

File: Readme.md [hunk 2]
```diff
@@ -88,9 +88,10 @@ app.use(convert(session(app)));
 ### Options
 
   The cookie name is controlled by the `key` option, which defaults
-  to "koa:sess". All other options are passed to `ctx.cookies.get()` and
-  `ctx.cookies.set()` allowing you to control security, domain, path,
-  and signing among other settings.
+  to "koa:sess". The cookie key is controlled by the `genid` option,
```
Comment (dead-horse) [inline Readme.md]: maybe we should put `genid` in external store section?

**Comment by dead-horse** [comment]
others LGTM

File: Readme.md (hunk 2 above)
Comment (shaoshuai0102) [inline Readme.md]: `genid` is designed for general purpose, not just work with external store. We could also change the way of generating id with cookie store with it.

Comment (dead-horse) [inline Readme.md]: no, `genid` is only work with external store, cookie store won't run into this logic. 

```js

# cookie store
koa:sess -> session vlaue
# external store
koa:sess -> genid() -> session value
```

Comment (shaoshuai0102) [inline Readme.md]: yeah, I am wrong. Thanks for reminding. I will change doc

**Comment by shaoshuai0102** [comment]
done

**Comment by dead-horse** [comment]
4.4.0

**Pull request merged**